esp/esdllib/README.md

```diff
@@ -0,0 +1,1608 @@
+```
+
+So far we've concluded that escaping is not required in these attributes:
+
+* `getDataSection()` function inside entry point `@source` attribute
+* `set-value[@value]` attribute
+
+### Implicit Variables
+
+These special-case variables are implicitly defined by the scripting engine. They give scripts access to characteristics of the current transaction and context. For example, say we've defined an ESP service named `WsPersonInfo` with a method named `GetPersonInfo` that has an interface defined as:
+
+```xml
+    <GetPersonInfoRequest>
+        <FirstName/>
+        <LastName/>
+    </GetPersonInfoRequest>
+```
+
+And is bound to a back end with this configuration:
+
+```xml
+    <Method name="GetPersonInfo"  url="https://backend.roxie.com:1234" mode="ESDL" queryname="PersonInfo_Services.GetPersonInfo" status="available"/>
+```
+
+* `$method`
+: The name of the ESP method invoked by the current transaction request. In our example this is the string *GetPersonInfo*.
+
+* `$request`
+: The name of the ESP request structure at the root of the current transaction request. In our example this is the string *GetPersonInfoRequest*.
+
+* `$service`
+: The name of the service containing the current transaction's method. In our example this is the string *WsPersonInfo*.
+
+* `$query`
+: The name of the back end roxie query. In our example this is the string *PersonInfo_Services.GetPersonInfo*.
+
+### System Parameters
+
+Each of these parameters gives the script access to some information about the current execution context. In order to reference a parameter, it must be declared in the script like so:
+
+```xml
+<es:param name="clientversion" select="'unknown'"/>
+```
+
+If the parameter is undefined, then it is assigned the value of the `@select` XPath expression.
+
+* `$clientversion` : The client interface version specified in the incoming request.
+* `$espUserName` : The username of the user authenticated to run the current transaction.
+* `$espUserPeer` : The IP address of host originating the current transaction.
+* `$espUserRealm` : The 'Company' or parent entity to the currently authenticated user.
+* `$espUserStatus` : The status of the currently authenticated user (eg *Active*, *Suspended*, etc., as defined by the security plugin and back end in use)
+
+In addition to the standard parameters above, any *Secure User Settings* configured as part of your service deployment are made available as parameters for you to declare and use in scripts. In configmgr, these settings are under ESP Process|ESP Service Bindings|Secure User Setting section. When configuring as an esdl application with a yaml file, it is in the authNZ.auth_map section
+
+### Variable Resolution
+Because of some implmentation details common implicit variables shouldn't be used in many XPath expressions as they aren't resolved correctly. For example you may want to extract a value from the incoming request using this command:
```

Comment:
It's not a bug, it's how xpath works.  You can't just put a variable in the path.

We could add a new feature where we allow substitution before compiling the xpath.  It would look more like this:

"soap:Body/{$query}/{$request}/Row/User/ReferenceCode".

XSLT supports that in some cases.  That's not xpath, that's preprocessing.

Reply:
I think the line should be rephrased.  To talk about how xpath doesn't allow variables to be used arbitrarily like that.

The problem with adding the preprocessor would likely be that we can't precompile the xpath anymore.  Not if those values are runtime variables.

Reply:
Don't mean to be discussing design here, but I could come up with a short hand for *[local-name()=$query] that is pre-processed.